// @sudoku/gameStore.js
/*
  Game Store Adapter - 将领域对象接入 Svelte 响应式系统
  
  职责：
  - 持有 Game / Sudoku 领域对象
  - 对外暴露 Svelte writable store 响应式状态
  - 对外暴露 UI 可调用的方法（guess, undo, redo 等）
  - 当领域对象变化时，自动更新 store 触发 UI 刷新
  
  设计说明：
  - 使用 Svelte writable store 作为桥接层
  - 领域对象变化时通过 subscribe 机制触发 store.update()
  - UI 组件通过 $store 语法消费响应式状态
  - gameStore.grid 成为 UI grid 数据的唯一来源
 */

import { writable } from 'svelte/store';
import { createGame, createSudoku, createGameFromJSON } from '@sudoku/domain';

export function createGameStore() {
  // 内部持有 Game 领域对象
  let game = null;

  // 响应式状态
  const grid = writable(Array(9).fill(null).map(() => Array(9).fill(0)));
  const invalidCells = writable([]);
  const canUndo = writable(false);
  const canRedo = writable(false);
  const isComplete = writable(false);

  // 从领域对象同步状态到 store
  function syncToStores() {
    if (!game) return;
    
    const sudoku = game.getSudoku();
    const currentGrid = sudoku.getGrid();
    grid.set(currentGrid);
    
    // 计算冲突单元格
    const invalid = sudoku.validate();
    invalidCells.set(invalid);
    
    canUndo.set(game.canUndo());
    canRedo.set(game.canRedo());
    isComplete.set(sudoku.isComplete());
  }

  return {
    // 响应式状态（可被 UI 订阅）
    grid,
    invalidCells,
    canUndo,
    canRedo,
    isComplete,

    // 初始化新游戏
    initGame(initialGrid) {
      const sudoku = createSudoku(initialGrid);
      game = createGame({ sudoku });
      
      // 订阅领域对象的变化
      game.subscribe(() => {
        syncToStores();
      });
      
      syncToStores();
    },

    // 获取当前 Game 对象（用于序列化等）
    getGame() {
      return game;
    },

    // 从 JSON 恢复游戏
    loadFromJSON(json) {
      game = createGameFromJSON(json);
      
      game.subscribe(() => {
        syncToStores();
      });
      
      syncToStores();
    },

    // 用户输入
    guess(row, col, value) {
      if (!game) return;
      game.guess({ row, col, value });
    },

    // 撤销
    undo() {
      if (!game) return;
      game.undo();
    },

    // 重做
    redo() {
      if (!game) return;
      game.redo();
    },

    // 检查是否可以撤销
    canUndo() {
      return game ? game.canUndo() : false;
    },

    // 检查是否可以重做
    canRedo() {
      return game ? game.canRedo() : false;
    }
  };
}

// 创建单例 game store
export const gameStore = createGameStore();
